### PR TITLE
Let users choose a share_folder

### LINKED ISSUE #4471: Set share_folder per user
Hello,

 I want to request the feature that the share_folder can be set per user (by the user) instead on server level.

**Current situation**:
In the moment, one can define in config.php via the value of share_folder where folders shared with a user are linked. For example if share_folder='/share' and user B shares folder DB with user A and user A shares folder DA with user B, then user A can access DB as /share/DB and user B can access DA as /share/DA. However per definition this can only defined on server level, i.e. for all users at the same time.

**Requested situation**:
I would like to have an option for every user (editable by the user), where this folder can be set. For example A could set 'his/her' share_folder='/share' and B 'his/her' share_folder='/' and therefore B can access DA in /share/DA while A can access DB in /DB.

Best regards,

Chris

### PR DESCRIPTION
**Pull request by Bibo-Joshi**
fix #4471

This PR aims at implementing the feature proposed in #4471, i.e. let users select their own `share_folder`.

I'm aware that this PR does not yet comply with the quality standards requested in the contribution guide (e.g. missing unit tests & unsigned commits) and I intend to fix that. However, as I'm new to web developement and especially haven't worked on NCs code base before, I'd like to first check if I'm headed in the right direction before I try to dig into the unit tests etc (tbh, the dev guide felt a bit overwhelming for a 1st time contributer :D).

### Some comments on this PR:
1. I did not commit the changed `.js` and `.map` because I wasn't sure if I have produced them entirely correctly (a lot of files changed that I didn't expect to) and this way it's also easier to get an overview of what I actually did
2. Because the structure of NC is new to me, I tried to make changes based on the surrounding/related code and there are probably more elegant solutions.
3. Currently admins can not forbid users from setting a custom `share_folder`.
4. I tested manually with both local & federated shares
5. For local shares the target location is computed when the share is created rather then when it's accepted. So when the recipient changes the `share_folder` between creation of the share and accepting it, it will be in the old folder. Changing this behavior seemed more intricate to me, so I let it slide for now.
6. Namings & wordings in the UI are just suggestions & I didn't look into i18n yet

### Questions that I have:
1. Is the general direction of what I did here okayish?
2.  Did I miss any places where the users setting should be taken into account? (Figuring out where shares are created & accepted wasn't too easy for me tbh …)
3. Would points 3 and 5 above have to be addressed for this PR to be taken into consideration?

I would be glad to get some comments :)

---

EDIT: I'm just now thinking of the case where a user might try to select a folder shared with them. This will probably lead to problems when they don't have write permission or are not allowed to re-share the incoming share. I'll try to have a closer look at these cases.

### CODE REVIEW

**Comment by szaimen** [comment]
Hi @Bibo-Joshi 👋,

Thanks a lot for your PR!

> 3\. Currently admins can not forbid users from setting a custom `share_folder`

Just wanted to add my thoughts, that I think this would be good having it implemented in this PR, too.

**Comment by Bibo-Joshi** [comment]
Added two things:

* Admins can now forbid users from using the new setting. I added no GUI option for this so far, but e.g. `'sharing.force_share_accept'` also doesn't have one …
* Added a button to quickly reset the custom value to the system default

Also regarding my edit about selecting a shared folder as `share_folder`: Apparently in this case, the shared folder is moved (`FolderName (2)`) and the `share_folder` gets the actual name, so no problems here. Would be nice if the directory picker could handle not displaying those folders in the first place, but tbh I haven't looked into that yet.

File: apps/files_sharing/appinfo/routes.php [hunk 1]
```diff
@@ -82,6 +82,16 @@
...
 		],
+		[
+			'name' => 'Settings#setShareFolder',
+			'url' => '/settings/shareFolder',
+			'verb' => 'PUT',
+		],
+		[
+			'name' => 'Settings#resetShareFolder',
+			'url' => '/settings/resetShareFolder',
+			'verb' => 'PUT',
```
Comment (skjnldsv) [inline apps/files_sharing/appinfo/routes.php:93]:
```suggestion
			'name' => 'Settings#resetShareFolder',
			'url' => '/settings/shareFolder',
			'verb' => 'DELETE',
```

Would fit the API more properly I think :)

**Comment by skjnldsv** [comment]
Rebased

**Comment by skjnldsv** [comment]
@jancborchardt @nimisha-vijay if you have a few time to suggest improvements on the settings ui, I would appreciate it! :hugs: 
image

**Comment by nimisha-vijay** [comment]
Looking good! Feedback from my side:
- Clicking on "Change" opens the folder picker, yeah? What if we instead removed the "Change" button and made the selected folder the one that opens the folder picker. We can make it look like an input field.
- "Reset" button will probably not be used much (I think?) so we can style it so that it's not as present (eg. only text with an underline)
- Same font size for "Accept user and group shares..." and "Set default folder..."

**Comment by skjnldsv** [comment]
Thanks for the feedback Nimisha! @Bibo-Joshi can you cover that? :)

> * "Reset" button will probably not be used much (I think?) so we can style it so that it's not as present (eg. only text with an underline)

We need to be able to go back to the system/instance default :)

**Review by skjnldsv — APPROVED** [review]
Other than the design fixes, all good from the technical pov :rocket:

**Comment by Bibo-Joshi** [comment]
> Thanks for the feedback Nimisha! @Bibo-Joshi can you cover that? :)

@skjnldsv I'll try! though it may be 1-2 weeks until I get around to it … Please let me now if you need me to do anything else, e.g. look into unit tests or i18n.

**Comment by Bibo-Joshi** [comment]
@skjnldsv I finally got around to it. Latest commit

* Make the folder selection an input field
* keeps the reset button
* makes sure that file picker opens the currently set share folder instead of the root

I also added myself to a couple of coypright headers.

**Comment by skjnldsv** [comment]
Hey!
I took the liberty of adapting this PR a bit to fit our more modern coding approach. That includes:
- Moving the user-folder functions into proper actions not using static methods: `setUserShareFolder` & `resetUserShareFolder`. That way it's a bit more straightforward to what it really does.
- I cleaned the vue component and made sure we use the html defaults (reset event)
- I implemented proper BEM-light structure for clarity as well as code documentation
- I implemented Nimisha's comment regarding the reset button:
  > "Reset" button will probably not be used much (I think?) so we can style it so that it's not as present (eg. only text with an underline)
- I made sure to only show the reset button when the folder is set to something different

dev skjnldsv com_settings_user_sharing (1)
dev skjnldsv com_settings_user_sharing

**Comment by Bibo-Joshi** [comment]
Thanks for putting in the effort @skjnldsv ! :)

Regarding the `Reset` action I probably misinterpreted your comment https://github.com/nextcloud/server/pull/28157#issuecomment-899571024 as "let's keep it as button". Personally, I think the underlined text is easily confused with a hyperlink, whereas a button clearly indicates that an action will take place. Also at quick glance, I couldn't find any other place in the UI, where hyperlinks are used instead of buttons. But that's just my 2cts …

Let me know if I can be of any more help with this PR :)

File: apps/files_sharing/src/components/SelectShareFolderDialogue.vue [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+<style scoped lang="scss">
+.share-folder {
+	&__form {
+		display: flex;
+	}
+
+	&__picker {
+		cursor: pointer;
+		min-width: 266px;
```
Comment (artonge) [inline apps/files_sharing/src/components/SelectShareFolderDialogue.vue:113]: Arbitrary?

**Review by artonge — CHANGES_REQUESTED** [review]

File: apps/files_sharing/src/components/SelectShareFolderDialogue.vue (hunk 2 above)
Comment (skjnldsv) [reply apps/files_sharing/src/components/SelectShareFolderDialogue.vue:113]: Totally :shrug: 
We don't really have a dedicated value to put here :thinking:

File: apps/files_sharing/src/components/PersonalSettings.vue [hunk 3]
```diff
@@ -38,23 +42,38 @@ import axios from '@nextcloud/axios'
...
 		toggleEnabled() {
-			axios.put(
-				generateUrl('/apps/files_sharing/settings/defaultAccept'),
-				{
-					accept: this.accepting,
-				}
-			)
+			axios.put(generateUrl('/apps/files_sharing/settings/defaultAccept'), {
+				accept: this.accepting,
+			})
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply apps/files_sharing/src/components/PersonalSettings.vue:68]: Yeah, that's not really related to this PR :sweat_smile:

File: apps/files_sharing/lib/External/Manager.php [hunk 4]
```diff
@@ -309,7 +309,7 @@ public function acceptShare($id) {
 
 		if ($share) {
 			\OC_Util::setupFS($this->uid);
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply apps/files_sharing/lib/External/Manager.php:311]: That's also unrelated to this PR, I would't risk touching this unknown area :thinking:

File: apps/files_sharing/src/components/SelectShareFolderDialogue.vue (hunk 2 above)
Comment (skjnldsv) [reply apps/files_sharing/src/components/SelectShareFolderDialogue.vue:113]: Kind of similar to the multiselect being set to 280 in another location: 
https://github.com/nextcloud/server/blob/6e5191b7021b448a48d741233bd17ab3c222119f/apps/files/src/components/TransferOwnershipDialogue.vue#L247

Any recommendations?

Comment (Bibo-Joshi) [reply apps/files_sharing/src/components/SelectShareFolderDialogue.vue:113]: yes, I tried to set it to the same width as the multiselect in the tranfer ownership thingy, but couldn't figure out where that was set. So I took the value that my browser had computed … 😬

**Review by artonge — APPROVED** [review]

**Comment by artonge** [comment]
Missing DCO, beside that, looks good :)

**Pull request merged**